### ISSUE TITLE
Move and Focus relative to left / right / up / down

### ISSUE THREAD

**Issue by adelin-b** [rickbutton/workspacer#206]
It is currently not possible to move or focus windows left / right / up / down

A config binding to something like `getWorkspace('up' | 'left' | 'right' | 'down ')` and  `getWindow('up' | 'left' | 'right' | 'down ')` would be really helpfulll

**Comment by system32uwu** [comment]
Alt+j and Alt+k are the default keybindings to cycle through the opened windows, is that what you are referring to by moving focus?
![image](https://user-images.githubusercontent.com/29718978/123563181-9ed23600-d789-11eb-9617-d1d24f97cae9.png)
[docs](https://workspacer.org/keybindings/)

**Comment by N1x0** [comment]
Workspaces:
Workspaces are represented as a list and hence and up/down movements don't really make sense to me. Maybe they could be used as fallback methods to just call the get_next and get_previous workspace functions but I am not sure if this will just cause more confusion. Please feedback.

Windows in a workspace:

At the moment workspacer only consider list indices when moving/focusing windows in a workspace. All LayoutEngines return an array/list that is then used to arrange windows. Because of this approach there is no simple way to implement the `up` and `down` navigation keys as workspacer currently does not check for the actual window position when focusing or moving windows.

One possible approach would be to implement a node/tree based layout similar to i3, but this would obviously require a bit of work.

Alternatively the up and down functions could check if the window intersects with the relevant virtualscreenborder and if it doesn't then check if the next window in the list is below the current one (win2.y > win1.y) . This would have to be done for all windows until the right one is found as there's no real guarantee that the windows are sorted appropriately.

If this is a heavily requested we should look into, would make it more userfriendly for those who come from a linux TWM background as well

**Comment by adelin-b** [comment]
Yes, checking the windows and finding the closest one on the top/down/left/right would be perfect.

But also for workspace, currently I have a 2*3 multi screen and to move up 1 workspace I have to go through 3 of them which are not even aligned (workspace 1 is on top, workspace 2 on bottom, workspace 3 on top etc)

The experience is so different than I3 or tiling windows manager on Linux that I went back to it (because a nice tiling wm is the last thing that prevent me to use window now)

**Comment by dalyIsaac** [comment]
We could probably add methods to allow focusing on specific workspaces and windows based on index. For example:

```csharp
// focus on the workspace if it is open in a monitor, otherwise swap focus to 
// it in the focused monitor
workspaces.Focus(int index);

// swap focus of the current monitor to the workspace
workspaces.SwapFocusTo(int index); 

workspaces.FocusedWorkspace.Focus(int index);
```

From that, you could set up keyboard shortcuts or add it to the menu. It's not exactly what you're asking for, but it's more in line with the current design. 

Would this solve your issue?

**Comment by adelin-b** [comment]
> We could probably add methods to allow focusing on specific workspaces and windows based on index. For example:
> 
> ```cs
> // focus on the workspace if it is open in a monitor, otherwise swap focus to 
> // it in the focused monitor
> workspaces.Focus(int index);
> 
> // swap focus of the current monitor to the workspace
> workspaces.SwapFocusTo(int index); 
> 
> workspaces.FocusedWorkspace.Focus(int index);
> ```
> 
> From that, you could set up keyboard shortcuts or add it to the menu. It's not exactly what you're asking for, but it's more in line with the current design.
> 
> Would this solve your issue?

I dont think it would works, Focusing the window on top or moving on the top screen would require to have an index of which screen in on top of which and then finding the corresponding index.  

If you can make this then it would work, but by itself just the index is not going to solve the issue.

**Comment by mattiasJohnson** [comment]
Are there any plans for implementing this or is there any plugin/config that does?
